dokka-subprojects/analysis-kotlin-descriptors-compiler/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/descriptors/compiler/impl/DescriptorSampleAnalysisEnvironment.kt

```diff
@@ -180,16 +193,90 @@ internal class DescriptorSampleAnalysisEnvironment(
             .trimIndent()
     }
 
-    private fun getSampleBody(sampleElement: PsiElement): String {
-        return when (sampleElement) {
+    private fun getSampleBody(psiElement: PsiElement): String {
+        return when (psiElement) {
             is KtDeclarationWithBody -> {
-                when (val bodyExpression = sampleElement.bodyExpression) {
-                    is KtBlockExpression -> bodyExpression.text.removeSurrounding("{", "}")
-                    else -> bodyExpression!!.text
+                val bodyExpression = psiElement.bodyExpression
+                val bodyExpressionText = bodyExpression!!.buildSampleText()
+                when (bodyExpression) {
+                    is KtBlockExpression -> bodyExpressionText.removeSurrounding("{", "}")
+                    else -> bodyExpressionText
                 }
             }
 
-            else -> sampleElement.text
+            else -> psiElement.text
         }
     }
+
+    private fun PsiElement.buildSampleText(): String {
+        val sampleBuilder = SampleBuilder(sampleRewriter)
+        this.accept(sampleBuilder)
+
+        sampleBuilder.errors.forEach {
+            val sw = StringWriter()
+            val pw = PrintWriter(sw)
+            it.e.printStackTrace(pw)
+
+            dokkaLogger.warn("${containingFile.name}: (${it.loc}): Exception thrown while converting \n```\n${it.text}\n```\n$sw")
+        }
+        return sampleBuilder.text
+    }
 }
+
+private class SampleBuilder(private val sampleRewriter: SampleRewriter?) : KtTreeVisitorVoid() {
```

Comment:
I am not sure I understand your idea correctly. Can you provide a constructor signature?


> Also, AFAIU, if SampleRewriter is null we don't even need SampleBuilder and should just use old logic, no?

Then we use the old logic.

Reply:
> Then we use the old logic.

It's a bit difficult to see that because the two paths (with sampleRewriter and without sampleRewriter) are handled in a single place. I think if the null check is extracted to `buildSampleText`, it'll be easier to understand:

```kotlin
private fun PsiElement.buildSampleText(): String {
    if (sampleRewriter == null) {
        return this.text
    }
    val sampleBuilder = SampleBuilder(sampleRewriter)
    ...
}
```

Reply:
I thought, that may be something like this will be better:
```kotlin
private fun PsiElement.buildSampleText(): String {
  if (sampleRewriter == null) return this.text

  val textBuilder = StringBuilder()
  val errors = mutableListOf<SampleConversionError>()

  this.accept(SampleBuilder(sampleRewriter, textBuilder, errors))

  errors.forEach { ... }
  return textBuilder.toString()
}
```

This way `SampleBuilder` will be just a visitor and all it's state will be provided outside, as it's consumed outside and only filled inside